### ISSUE TITLE
Failure to revert function pointers in `json_set_allocation_functions` in Parson

### ISSUE THREAD

**Issue by mwhicks1** [correctcomputation/C3#5]
The [checkedc-parson benchmark](https://github.com/correctcomputation/C3/tree/master/samples/checkedc-parson) does not compile (there's an [open issue about this](https://github.com/microsoft/checkedc-parson/issues/20)).

That said, it does unconvert properly with C3 (after editing it slightly to deal with C3's limitations), and that version compiles and runs fine. We should fix the original Checked C so that it works.

**Comment by mwhicks1** [comment]
@aaronjeline got the original Checked C to compile, but C3 is not able to revert it. I extended the `fptr.c` test case with this bit:
```
_Itype_for_any(T) void json_set_allocation_functions(_Ptr<void* (size_t s) : itype(_Array_ptr<T>) byte_count(s)> malloc,
    _Ptr<void (void* : itype(_Array_ptr<T>) byte_count(0))> free);
```
which gets hung up in the function pointer somewhere.

**Comment by aaronjeline** [comment]
More minimal test case: `_Ptr<void (*) (int)> x = 0`

Correct conversion: `void (**x) (int) = 0;`

Will be a little weird to move the `*` into the middle

**Comment by mattmccutchen-cci** [comment]
After discussion with Aaron, there are three different test cases above:

1. Pointer to a function that returns a pointer (`malloc` parameter of `json_set_allocation_functions`)
2. Pointer to a function that takes a pointer as a parameter (`free` parameter of `json_set_allocation_functions`)
3. Pointer to a pointer to a function as in Aaron's comment.

Currently, C3 can't parse any of these.  The first two are the ones we currently care about for Parson.

**Comment by mattmccutchen-cci** [comment]
I was messing around to learn how C3 works, and I came up with #13, which gives the answer I think we want for `json_set_allocation_functions`, though probably not via an acceptable approach.  I don't know what I'm doing here, and I've asked Aaron to help me learn.

One thing I think is worth adding to the description of the test cases above: The fact that the parameter names of `json_set_allocation_functions` are `malloc` and `free` seems to be part of the problem, because C3 lexes those names as a special token type `PID` that isn't recognized everywhere that `ID` is.

**Comment by aaronjeline** [comment]
I tested @mattmccutchen-cci's commit against all the `samples` at it looks like it doesn't break anything.

**Comment by mwhicks1** [comment]
Fixed by #13
